File: c/stack.c
#include <windows.h>
#include <stdio.h>


typedef struct Node_
{
    unsigned int val;
    struct Node_* bot;
} Node;

typedef struct Stack_
{
    Node* top;
} Stack;

void push_stack(Stack* stack, int a)
{
    Node* node = (Node*)malloc(sizeof(Node));
    node->val = a;
    node->bot = stack->top;
    stack->top = node;
}

unsigned int pop_stack(Stack* stack)
{
    if (!stack->top)
    {
        printf("Stack is emtpy\n");
        return 0;
    }

    Node* tmp = stack->top;
    int out = tmp->val;
    stack->top = tmp->bot;
    free(tmp);
    return out;
}

int peek_stack(Stack* stack)
{
    if (!stack->top)
    {
        printf("stack is empty\n");
        return 0;
    }
    return stack->top->val;
}

void print_stack(Stack* stack)
{
    if (!stack->top)
    {
        printf("Stack is empty\n");
        return;
    }

    printf("%u <- top\n", stack->top->val);

    Node* node = stack->top->bot;
    while (node)
    {
        printf("%u\n", node->val);
        node = node->bot;
    }
}

void free_stack(Stack* stack)
{
    while (stack->top)
    {
        struct Node* t = stack->top;
        stack->top = stack->top->bot;
        free(t);
    }
}

int main(void)
{
    Stack stack = { NULL };

    push_stack(&stack, 400);
    push_stack(&stack, 500);
    push_stack(&stack, 600);
    push_stack(&stack, 700);

    print_stack(&stack);

    printf("\n");

    printf("popped: %u\n\n", pop_stack(&stack));

    print_stack(&stack);

    free_stack(&stack);
}

File: python/stack.py


class Node:
    def __init__(self, x):
        self.val = x
        self.bot = None

class Stack:
    def __init__(self):
        self.top = None

    def empty(self) -> bool:
        return self.top is None

    def push(self, x):
        node = Node(x)  
        node.bot = self.top
        self.top = node
    
    def pop(self) -> int:
        if self.empty():
            print("stack empty")
            return
        tmp = self.top
        val = tmp.val
        self.top = tmp.bot
        return val

    def print_all(self):
        print(f"{self.top.val} <- top")
        node = self.top.bot
        while node:
            print(f"{node.val}")
            node = node.bot
    def peek(self):
        if self.empty():
            print("Stack empty")
            return None
        return self.top.val
    
stack = Stack()

stack.push(400)
stack.push(500)
stack.push(600)
stack.print_all()
print()

y = stack.pop()
stack.print_all()
print()

